xdp/src/netlink.rs

```diff
@@ -0,0 +1,601 @@
+            continue;
+        };
+
+        neighbors.push(neighbor);
+    }
+
+    Ok(neighbors)
+}
+
+pub fn parse_rtm_newneigh(msg: NetlinkMessage, if_index: Option<i32>) -> Option<NeighborEntry> {
+    let nd_msg = unsafe { ptr::read_unaligned(msg.data.as_ptr() as *const ndmsg) };
+    if let Some(idx) = if_index {
+        if nd_msg.ndm_ifindex != idx {
+            return None;
+        }
+    }
+    let Ok(attrs) = parse_attrs(&msg.data[mem::size_of::<ndmsg>()..]) else {
+        return None;
+    };
+    let mut neighbor = NeighborEntry {
+        destination: None,
+        lladdr: None,
+        ifindex: nd_msg.ndm_ifindex,
+        state: nd_msg.ndm_state,
+    };
+    if let Some(dst_attr) = attrs.get(&NDA_DST) {
+        neighbor.destination = parse_ip_address(dst_attr.data, nd_msg.ndm_family);
+    }
+    if let Some(lladdr_attr) = attrs.get(&NDA_LLADDR) {
+        if lladdr_attr.data.len() >= 6 {
+            let mut mac = [0u8; 6];
+            mac.copy_from_slice(&lladdr_attr.data[0..6]);
+            neighbor.lladdr = Some(MacAddress(mac));
+        }
+    }
+    Some(neighbor)
+}
+
+#[derive(Debug, Clone)]
+pub struct RouteEntry {
+    pub destination: Option<IpAddr>,
+    pub gateway: Option<IpAddr>,
+    pub pref_src: Option<IpAddr>,
+    pub out_if_index: Option<i32>,
+    pub in_if_index: Option<i32>,
+    pub priority: Option<u32>,
+    pub table: Option<u32>,
+    pub protocol: u8,
+    pub scope: u8,
+    pub type_: u8,
+    pub family: u8,
+    pub dst_len: u8,
+}
+
+#[repr(C)]
+#[allow(non_camel_case_types)]
+struct rtmsg {
+    rtm_family: u8,
+    rtm_dst_len: u8,
+    rtm_src_len: u8,
+    rtm_tos: u8,
+    rtm_table: u8,
+    rtm_protocol: u8,
+    rtm_scope: u8,
+    rtm_type: u8,
+    rtm_flags: u32,
+}
+
+#[repr(C)]
+struct RouteRequest {
+    header: nlmsghdr,
+    rtm: rtmsg,
+}
+
+fn parse_ip_address(data: &[u8], family: u8) -> Option<IpAddr> {
+    match family as i32 {
+        AF_INET if data.len() == 4 => Some(IpAddr::V4(Ipv4Addr::new(
+            data[0], data[1], data[2], data[3],
+        ))),
+        AF_INET6 if data.len() == 16 => {
+            let mut segments = [0u16; 8];
+            for i in 0..8 {
+                segments[i] = ((data[i * 2] as u16) << 8) | (data[i * 2 + 1] as u16);
+            }
+            Some(IpAddr::V6(Ipv6Addr::from(segments)))
+        }
+        _ => None,
+    }
+}
+
+pub fn netlink_get_routes(family: u8) -> Result<Vec<RouteEntry>, io::Error> {
+    let sock = NetlinkSocket::open()?;
+
+    // Safety: RouteRequest is POD
+    let mut req = unsafe { mem::zeroed::<RouteRequest>() };
+
+    let nlmsg_len = mem::size_of::<nlmsghdr>() + mem::size_of::<rtmsg>();
+    req.header = nlmsghdr {
+        nlmsg_len: nlmsg_len as u32,
+        nlmsg_flags: (NLM_F_REQUEST | NLM_F_DUMP) as u16,
+        nlmsg_type: RTM_GETROUTE,
+        nlmsg_pid: 0,
+        nlmsg_seq: 1,
+    };
+
+    req.rtm.rtm_family = family;
+    req.rtm.rtm_table = RT_TABLE_MAIN;
```

Comment:
We will want to make the routing table ID configurable for future compatibility with DZ

Reply:
I don't plan to do this in this PR. We can write code for the future in the future.